File: ruby-problems/MinStack.rb
# source: LeetCode
# https://leetcode.com/problems/min-stack/

# 155. <easy> MinStack
=begin
Design a stack that supports push, pop, top, and retrieving the minimum element in constant time.

    push(x) -- Push element x onto stack.
    pop() -- Removes the element on top of the stack.
    top() -- Get the top element.
    getMin() -- Retrieve the minimum element in the stack.



Example 1:

Input
["MinStack","push","push","push","getMin","pop","top","getMin"]
[[],[-2],[0],[-3],[],[],[],[]]

Output
[null,null,null,null,-3,null,0,-2]

Explanation
MinStack minStack = new MinStack();
minStack.push(-2);
minStack.push(0);
minStack.push(-3);
minStack.getMin(); // return -3
minStack.pop();
minStack.top();    // return 0
minStack.getMin(); // return -2



Constraints:

    Methods pop, top and getMin operations will always be called on non-empty stacks.
=end

class MinStack

    private
    attr_accessor :stack, :min_stack


    public

=begin
    initialize your data structure here.
=end
    def initialize()
        @stack = []
        @min_stack = []
    end


=begin
    :type x: Integer
    :rtype: Void
=end
    def push(x)
        evaluate_min_push(x)
        stack.push(x)
    end


=begin
    :rtype: Void
=end
    def pop()
        evaluate_min_pop()
        stack.pop
    end


=begin
    :rtype: Integer
=end
    def top()
        return stack.last
    end


=begin
    :rtype: Integer
=end
    def get_min()
        return min_stack.last
    end


    private

=begin
    :type x: Integer
    :rtype: Void
=end
    def evaluate_min_push(x)
        current_min = min_stack.last
        if current_min.nil? || x <= current_min
            min_stack.push(x)
        end
    end

=begin
    :rtype: Void
=end
    def evaluate_min_pop()
        last_item = stack.last
        current_min = min_stack.last
        if last_item == current_min
            min_stack.pop
        end
    end

end

# Your MinStack object will be instantiated and called as such:
# obj = MinStack.new()
# obj.push(x)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.get_min()

=begin
18 / 18 test cases passed.
Runtime: 48 ms
Memory Usage: 12.3 MB
=end

File: java-problems/MinStack.java
// source: LeetCode
// https://leetcode.com/problems/min-stack

// 155. <easy> Min Stack
/*
Design a stack that supports push, pop, top, and retrieving the minimum element in constant time.

    push(x) -- Push element x onto stack.
    pop() -- Removes the element on top of the stack.
    top() -- Get the top element.
    getMin() -- Retrieve the minimum element in the stack.

Example 1:

Input
["MinStack","push","push","push","getMin","pop","top","getMin"]
[[],[-2],[0],[-3],[],[],[],[]]

Output
[null,null,null,null,-3,null,0,-2]

Explanation
MinStack minStack = new MinStack();
minStack.push(-2);
minStack.push(0);
minStack.push(-3);
minStack.getMin(); // return -3
minStack.pop();
minStack.top();    // return 0
minStack.getMin(); // return -2

Constraints:

    Methods pop, top and getMin operations will always be called on non-empty stacks.
*/

import java.util.Deque;
import java.util.ArrayDeque;

class MinStack {
    private Deque<Integer> stack;
    private Deque<Integer> minStack;

    /** initialize your data structure here. */
    public MinStack() {
        stack = new ArrayDeque<Integer>();
        minStack = new ArrayDeque<Integer>();
    }

    public void push(int x) {
        // add to stack
        stack.addFirst(x);
        // add to minStack if minStack is empty or x < current min
        Integer currMin = minStack.peekFirst();
        if (currMin == null || x <= currMin) {
            minStack.addFirst(x);
        }
    }

    public void pop() {
        // pop off stack
        Integer pop = stack.removeFirst();
        // pop off minStack if item from stack is on top of the minStack
        Integer currMin = minStack.peekFirst();
        if (currMin != null && currMin.equals(pop)) {
            minStack.removeFirst();
        }
    }

    public int top() {
        Integer top = stack.peekFirst();
        return top;
    }

    public int getMin() {
        Integer min = minStack.peekFirst();
        return min;
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(x);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */

 /*
 18 / 18 test cases passed.
Runtime: 4 ms
Memory Usage: 40.8 MB
 */

